### PR TITLE
ARROW-15192: [Java] Allow use of Jackson 2.12 and higher

### PR DESCRIPTION
**Pull request by laurentgo**
Jackson 2.12 does not enable by default (de)serializers for Java
Date/Time java.time.* classes and require the jsr310 module to be added
to the object mapper. The absence of this module causes vector module to
stop working with newer versions of Jackson.

Address this by adding a ObjectMapper factory class which register the
jsr310 module with the objectmapper used by vector classes.

### CODE REVIEW

File: java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java
Comment (Jimexist) [inline java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java:36]: can you add a unit test for how this will alter the behavior?

File: java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java
Comment (Jimexist) [inline java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java:32]:
```suggestion
   * Creates a new {@link ObjectMapper} instance preconfigured with Java time handling.
```

File: java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java
Comment (laurentgo) [reply java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java:36]: That's the tricky part, it does not alter the the current behavior, but if you want to convince yourself this patch works, you can try to build arrow with `-Ddep.jackson.version=2.12.6` without and with the patch to observe the issue is being addressed.

File: java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java
Comment (laurentgo) [reply java/vector/src/main/java/org/apache/arrow/vector/util/ObjectMapperFactory.java:32]: Isn't that an implementation detail though?

**Comment by laurentgo** [comment]
@Jimexist any other comments?

**Review by jimexist — APPROVED** [review]

**Comment by Jimexist** [comment]
i havent tested this locally - so if anyone would give a second opinion i'll wait for a few days before trying to merge as is

**Comment by laurentgo** [comment]
@Jimexist sorry to put some pressure on you, but since you approved this 3weeks ago, is it possible to submit this one?

**Review by lidavidm — APPROVED** [review]
I can confirm current HEAD fails unit tests with `-Ddep.jackson.version=2.12.6` but passes them with the patch. We should update the Jackson version in general, see ARROW-16143

**Pull request closed without merging**
